File: uvc.py
import v4l2
import numpy as np
import cv2
import sys
import time
import os
import threading

import RPi.GPIO as GPIO

# os.system('v4l2-ctl --set-fmt-video=width=640,height=480,pixelformat=GREY')

# User Controls
exposure = 0x00980911  # (int) : min=4 max=906 step=1 default=800 value=800
horizontal_flip = 0x00980914  # (bool): default=0 value=0
vertical_flip = 0x00980915  # (bool): default=0 value=0

# (int) : min=0 max=32765 step=1 default=-8193 value=240
brightness = 0x00980900
gain = 0x00980913  # (int): min=16 max=64 step=1 default=57343 value=32

# Camera Controls
# (menu):min=0 max=2 default=2 value=2 flags=read-only
camera_orientation = 0x009a0922
# (int):min=0 max=0 step=1 default=0 value=0 flags=read-only
camera_sensor_rotation = 0x009a0923

# Image Source Controls
# (int) : min=21 max=32367 step=1 default=21 value=21
vertical_blanking = 0x009e0901
# (int) : min=816 max=816 step=1 default=816 value=816 flags=read-only
horizontal_blanking = 0x009e0902
analogue_gain = 0x009e0903  # (int) : min=16 max=248 step=1 default=16 value=16

# Image Processing Controls
# (intmenu): min=0 max=0 default=0 value=0 flags=read-only
link_frequency = 0x009f0901
# (int64)  : min=200000000 max=200000000 step=1 default=200000000 value=2000000
pixel_rate = 0x009f0902


def thread_show():
    global g_image
    global image
    global semaphore
    while True:
        semaphore.acquire()

        # lock.acquire()
        # image_show = g_image
        # lock.release()

        cv2.imshow('image', g_image)

        key = cv2.waitKey(1)
        if key == ord('q'):
            sys.exit()


def main():
    global image
    global g_image
    global semaphore

    t1 = threading.Thread(target=thread_show)

    camera = v4l2.open2('/dev/video2', width, height, 'YUYV')
    v4l2.start(camera)
    # v4l2.setformat(camera,width,height,'GREY')

    # v4l2.setcontrol(camera, exposure, 2900)
    # value = v4l2.getcontrol(camera, exposure)

    v4l2.setcontrol(camera, brightness, 240)
    v4l2.setcontrol(camera, gain, 32)

    # v4l2.setcontrol(camera, V4L2_CID_HBLANK, 800)
    # value = v4l2.getcontrol(camera,V4L2_CID_HBLANK)

    # value = v4l2.getcontrol(camera, vertical_blanking)
    # v4l2.setcontrol(camera, vertical_blanking, 3000)
    # value = v4l2.getcontrol(camera, vertical_blanking)

    t1.start()

    frame_count = 0

    fps_count = 0
    fps = 0
    ts = time.time()

    # while frame_count < 1000:
    while True:
        GPIO.output(17, False)
        GPIO.output(18, False)

        data = v4l2.read(camera)

        GPIO.output(17, True)
        GPIO.output(18, True)

        # time.sleep(0.005)

        fps_count += 1
        if fps_count >= 30:
            t = time.time() - ts
            fps = int(round(fps_count / t, 0))
            print("fps:%d   frame length:%d" % (fps, len(data)))
            fps_count = 0
            ts = time.time()

        frame_count += 1
        if frame_count >= 1:
            frame_count = 0

            image_array = np.frombuffer(data, dtype=np.uint8)
            array0 = image_array[0::2]
            array1 = image_array[1::2]
            image0 = array0.reshape(height, width)
            image1 = array1.reshape(height, width)
            # image = np.hstack((image0, image1))
            image = image0

            # image = cv2.resize(image, (960, 360))
            # image = cv2.flip(image, 0)

            image = cv2.cvtColor(image, cv2.COLOR_GRAY2BGR)
            image = cv2.putText(image, 'FPS:'+str(fps), (10, 30),
                                cv2.FONT_HERSHEY_SIMPLEX, 0.5, (80, 80, 240), 1)

            lock.acquire()
            g_image = image
            lock.release()
            semaphore.release()

        cv2.imshow('image', image)
        key = cv2.waitKey(1)
        if key == ord('q'):
            sys.exit()

    v4l2.stop(camera)
    v4l2.close(camera)


gpio_state = True
GPIO.setmode(GPIO.BCM)
GPIO.setup(17, GPIO.OUT)
GPIO.setup(18, GPIO.OUT)

# width, height = 1280, 800
# width, height = 1280, 720
width, height = 640, 480
g_image = np.zeros((height, width), dtype=np.uint8)
lock = threading.Lock()
semaphore = threading.Semaphore(0)

if __name__ == '__main__':
    main()

File: uvc.cpp
#include "v4l2.hpp"

#include <sys/time.h>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/opencv.hpp>

using namespace cv; // 省去函数前面加cv::的必要性

static timeval t0;
static timeval t1;

static timeval brenner_t0;
static timeval brenner_t1;

static char text[255];

#pragma GCC push_options
#pragma GCC optimize("O3")
double brenner(cv::Mat &image)
{
    // assert(image.empty());

    // cv::Mat gray_img;
    // if (image.channels() == 3)
    // {
    //     cv::cvtColor(image, gray_img, CV_BGR2GRAY);
    // }
    // else
    // {
    //     gray_img = image;
    // }

    double result = .0f;
    for (int i = 0; i < image.rows; ++i)
    {
        uchar *data = image.ptr<uchar>(i);
        for (int j = 0; j < image.cols - 2; ++j)
        {
            result += pow(data[j + 2] - data[j], 2);
        }
    }

    return result;
    // return result / image.total();
}
#pragma GCC pop_options

// float brenner(Mat img, int width, int height)
// {
//     // '''
//     // :param img:narray 二维灰度图像
//     // :return: float 图像越清晰越大
//     // '''

//     uchar *ptmp = NULL;
//     float out = 0;
//     for (int x = 0; x < height - 2; x++)
//     {
//         for (int y = 0; y < width; y++)
//         {
//             out += ((int)(img.data[x + 2][y]) - (int)(img.data[x][y])) ^ 2;
//         }
//     }

//     return out;
// }

// #define WIDTH 640
// #define HEIGHT 400
#define WIDTH 640
#define HEIGHT 480

__u8 buffer[WIDTH * HEIGHT];

int main(int argc, char **argv)
{
    __u16 regval;
    struct buffer data;

    // void *handle = open("/dev/video0");
    void *handle = open("/dev/video2", WIDTH, HEIGHT, "YUYV"); //ssjhs GREY, not GRAY!!!!!!!!!!

    // setformat(handle, 1280,800,'GREY');

    getcontrol(handle, V4L2_CID_BRIGHTNESS);
    setcontrol(handle, V4L2_CID_BRIGHTNESS, 240);
    setcontrol(handle, V4L2_CID_GAIN, 32);

    regval = get_sensor_reg(handle, 0x0b);
    printf("get_sensor_reg:%d \r\n", regval);

    set_sensor_reg(handle, 0x0b, 300);
    regval = get_sensor_reg(handle, 0x0b);
    printf("get_sensor_reg:%d \r\n", regval);

    start(handle);

    gettimeofday(&t0, NULL);
    float t;
    int fps_count = 0;
    int fps;
    // double value = 0;
    // float interval;

    Mat img = Mat::zeros(HEIGHT, WIDTH, CV_8U);
    // Mat dst = Mat::zeros(480, 768, CV_8UC1);
    // Mat dst = Mat::zeros(600, 960, CV_8UC1);

    for (;;)
    {
        // snapshot(handle);
        data = read(handle);

        __u8* pointer = (__u8*)data.start;

// #pragma GCC push_options
// #pragma GCC optimize("O1")
        for (unsigned int i = 0; i < sizeof(buffer); i++)
        {
            buffer[i] = pointer[i * 2];
        }
// #pragma GCC pop_options

        Mat frame(HEIGHT, WIDTH, CV_8UC1, (unsigned char *)buffer);

        flip(frame, img, 0);
        // resize(img, dst, dst.size(), 0, 0, INTER_LINEAR);

        fps_count += 1;
        if (fps_count >= 30)
        {
            gettimeofday(&t1, NULL);
            t = (t1.tv_sec - t0.tv_sec) * 1000 + (t1.tv_usec - t0.tv_usec) / 1000;
            fps = (int)(fps_count * 1000 / t);
            sprintf(text, "FPS: %d", fps);
            printf("FPS: %d\n", fps);
            // printf("brenner: %f\n", value);
            // printf("interval: %f\n", interval);
            fps_count = 0;
            gettimeofday(&t0, NULL);
        }

        // gettimeofday(&brenner_t0, NULL);
        // value = brenner(img);
        // gettimeofday(&brenner_t1, NULL);
        // interval = (brenner_t1.tv_sec - brenner_t0.tv_sec) * 1000 + (brenner_t1.tv_usec - brenner_t0.tv_usec) / 1000;

        putText(img, text, cv::Point(10, 30), FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(255, 255, 0), 1); //FONT_HERSHEY_SIMPLEX

        imshow("image", img);
        waitKey(1);
    }

    stop(handle);
    close(handle);

    fprintf(stderr, "\n");

    return 0;
}
